expression/builtin_json_test.go

```diff
@@ -379,60 +379,73 @@ func (s *testEvaluatorSuite) TestJSONContains(c *C) {
-		{[]interface{}{`{"a": [1, 2, {"aa": "xx"}]}`, `1`, "$[*]"}, nil, false},
-		{[]interface{}{`{"a": [1, 2, {"aa": "xx"}]}`, `1`, "$**.a"}, nil, false},
+		{[]interface{}{`{"a": [1, 2, {"aa": "xx"}]}`, `1`, "$.*"}, nil, "[json:3149]In this situation, path expressions may not contain the * and ** tokens."},
+		{[]interface{}{`{"a": [1, 2, {"aa": "xx"}]}`, `1`, "$[*]"}, nil, "[json:3149]In this situation, path expressions may not contain the * and ** tokens."},
+		{[]interface{}{`{"a": [1, 2, {"aa": "xx"}]}`, `1`, "$**.a"}, nil, "[json:3149]In this situation, path expressions may not contain the * and ** tokens."},
 		// Tests path expression does not identify a section of the target document
-		{[]interface{}{`{"a": [1, 2, {"aa": "xx"}]}`, `1`, "$.c"}, nil, true},
-		{[]interface{}{`{"a": [1, 2, {"aa": "xx"}]}`, `1`, "$.a[3]"}, nil, true},
-		{[]interface{}{`{"a": [1, 2, {"aa": "xx"}]}`, `1`, "$.a[2].b"}, nil, true},
+		{[]interface{}{`{"a": [1, 2, {"aa": "xx"}]}`, `1`, "$.c"}, nil, ""},
+		{[]interface{}{`{"a": [1, 2, {"aa": "xx"}]}`, `1`, "$.a[3]"}, nil, ""},
+		{[]interface{}{`{"a": [1, 2, {"aa": "xx"}]}`, `1`, "$.a[2].b"}, nil, ""},
+		// For issue 9957: test 'argument 1 and 2 as valid json object'
+		{[]interface{}{`[1,2,[1,3]]`, `a:1`}, 1, "[json:3140]Invalid JSON text: invalid character 'a' looking for beginning of value"},
+		{[]interface{}{`a:1`, `1`}, 1, "[json:3140]Invalid JSON text: invalid character 'a' looking for beginning of value"},
 	}
 	for _, t := range tbl {
 		args := types.MakeDatums(t.input...)
 		f, err := fc.getFunction(s.ctx, s.datumsToConstants(args))
 		c.Assert(err, IsNil)
 		d, err := evalBuiltinFunc(f, chunk.Row{})
-		if t.success {
+		if t.errMsg == "" {
 			c.Assert(err, IsNil)
 			if t.expected == nil {
 				c.Assert(d.IsNull(), IsTrue)
 			} else {
 				c.Assert(d.GetInt64(), Equals, int64(t.expected.(int)))
 			}
 		} else {
-			c.Assert(err, NotNil)
+			c.Assert(err.Error(), Equals, t.errMsg)
 		}
 	}
+	// For issue 9957: test 'argument 1 and 2 as valid json object'
+	_, err := fc.getFunction(s.ctx, s.datumsToConstants(types.MakeDatums(1, ``)))
+	c.Assert(err.Error(), Equals, "[json:3146]Invalid data type for JSON data in argument 1 to function json_contains; a JSON string or JSON type is required.")
+	_, err = fc.getFunction(s.ctx, s.datumsToConstants(types.MakeDatums(0.05, ``)))
+	c.Assert(err.Error(), Equals, "[json:3146]Invalid data type for JSON data in argument 1 to function json_contains; a JSON string or JSON type is required.")
+	_, err = fc.getFunction(s.ctx, s.datumsToConstants(types.MakeDatums(``, 1)))
+	c.Assert(err.Error(), Equals, "[json:3146]Invalid data type for JSON data in argument 2 to function json_contains; a JSON string or JSON type is required.")
+	_, err = fc.getFunction(s.ctx, s.datumsToConstants(types.MakeDatums(``, 0.05)))
+	c.Assert(err.Error(), Equals, "[json:3146]Invalid data type for JSON data in argument 2 to function json_contains; a JSON string or JSON type is required.")
```

Comment:
Can we move these tests to tbl?

Reply:
@erjiaqing Those errors are returned by `getFunctions`, IMHO trying to integrate into `tbl` will make the codes complicated and confusing.

Reply:
Can we check them in line 427?

Reply:
Or, at least, use a loop.

Reply:
> Can we check them in line 427?
That means `tbl` need 2 error fields, which is ugly IMHO

Reply:
> Or, at least, use a loop.

addressed, thanks!